ocaml/xapi/repository.ml

```diff
@@ -0,0 +1,116 @@
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module Unixext = Xapi_stdext_unix.Unixext
+
+module D = Debug.Make (struct let name = "repository" end)
+
+open D
+
+let reposync_mutex = Mutex.create ()
+
+let create_repository_record ~__context ~name_label ~name_description ~binary_url ~source_url =
+  let ref = Ref.make () in
+  let uuid = Uuidm.to_string (Uuidm.create `V4) in
+  Db.Repository.create
+    ~__context
+    ~ref
+    ~uuid
+    ~name_label
+    ~name_description
+    ~binary_url
+    ~source_url
+    ~hash:""
+    ~up_to_date:false;
+  ref
+
+let assert_url_is_valid ~url =
+  try
+    let uri = Uri.of_string url in
+    match Uri.scheme uri with
+    | Some "http" | Some "https" -> 
+      begin match Uri.host uri with
+        | Some h ->
+          begin match !Xapi_globs.repository_domain_name_allowlist with
+            | [] -> ()
+            | l ->
+              begin match List.exists (fun d -> Astring.String.is_suffix ("." ^ d) h) l with
+                | true -> ()
+                | false ->
+                  let msg = "host is not in allowlist" in
+                  raise Api_errors.(Server_error (internal_error, [msg]))
+              end
+          end
+        | None -> raise Api_errors.(Server_error (internal_error, ["invalid host in url"]))
+      end
+    | _ -> raise Api_errors.(Server_error (internal_error, ["invalid scheme in url"]))
+  with e ->
+    error "Invalid url %s: %s" url (ExnHelper.string_of_exn e);
+    raise Api_errors.(Server_error (invalid_base_url, [url]))
+
+let introduce ~__context ~name_label ~name_description ~binary_url ~source_url =
+  assert_url_is_valid ~url:binary_url;
+  assert_url_is_valid ~url:source_url;
+  Db.Repository.get_all ~__context
+  |> List.iter (fun ref ->
+      if name_label = Db.Repository.get_name_label ~__context ~self:ref
+      || binary_url = Db.Repository.get_binary_url ~__context ~self:ref then
+        raise Api_errors.( Server_error (repository_already_exists, [(Ref.string_of ref)]) ));
+  create_repository_record  ~__context ~name_label ~name_description ~binary_url ~source_url
+
+let forget ~__context ~self =
+  let pool = Helpers.get_pool ~__context in
+  let enabled = Db.Pool.get_repository ~__context ~self:pool in
+  if enabled = self then
+    raise Api_errors.(Server_error (repository_is_in_use, []))
+  else
+    Db.Repository.destroy ~__context ~self
+
+let with_reposync_lock f =
+  if Mutex.try_lock reposync_mutex then
+    Xapi_stdext_pervasives.Pervasiveext.finally
```

Comment:
It's probably better to use this finally equivalent from the stdlib: `Fun.protect ~finally:(fun () -> Mutex.unlock reposync_mutex) f `

Reply:
I don't think the stdext's `finally` is equivalent to `Fun.protect`, they handle some exceptions differently. Let me find the github thread where these were talked about.

Reply:
I looked at this for a while. It looks like that the `Xapi_stdext_pervasives.Pervasiveext.finally` has an extra feature "Backtrace.is_important exn".
To be honest, I don't clearly understand the function of this: https://github.com/xapi-project/backtrace/blob/master/lib/backtrace.ml#L195

But as this `finally` is widely used in Toolstack, I think it would be fine to use it here.

Reply:
The difference is `Fun.protect` does not try to do any work when the finalizing function throws an exception and covers the original exception. The ocaml designers consider this a programming error.

`Pervasives.finally` protects against these bugs and logs the exception raised by the finaliser.